File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-0.java
class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        int[][] dp = new int[matrix.length][matrix[0].length];
        int[][] directions = {{0,1},{0,-1},{1,0},{-1,0}};
        int max = 0;
        for (int i =0;i<matrix.length;i++){
            for(int j =0;j<matrix[0].length;j++){
                dp[i][j] = dfs(dp,i,j,matrix,-1,directions);
                max = Math.max(max, dp[i][j]);
            }
        }
        return max;
    }
    public int dfs(int[][] dp, int row, int col, int[][]matrix, int prev, int[][]directions){
        if (row<0||col<0||row>=matrix.length||col>=matrix[0].length|| prev >=matrix[row][col]){
            return 0;
        }
        if(dp[row][col]!=0){
            return dp[row][col];
        }
        int max = 1;
        for (int[] dir : directions){
            max = Math.max(max, 1+ dfs(dp, row + dir[0], col+dir[1], matrix, matrix[row][col], directions));
        }
        dp[row][col] = max;
        return max;

    }
}

File: Data Structures & Algorithms/surrounded-regions/submission-0.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        
        directions =[[0,1],[0,-1],[-1,0],[1,0]]
        def dfs(i,j, board, letter1, letter2):
            if i < 0 or j < 0 or i >=len(board) or j >=len(board[0]) or board[i][j]!=letter1:
                return
            board[i][j]=letter2
            for row, col in directions:
                new_row = i + row
                new_col = j + col
                dfs(new_row,new_col,board, letter1, letter2)
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if i==0 or j ==0 or i ==len(board)-1 or j==len(board[0])-1:
                    dfs(i,j,board,'O','B')
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                dfs(i,j,board,'O','X')
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                dfs(i,j,board,'B','O')
        